File: contracts/tests/test_owner_count.py
# ethereum
from ethereum import tester as t
from ethereum.utils import sha3, privtoaddr, to_string
from ethereum.tester import ContractCreationFailed
# standard libraries
from unittest import TestCase


class TestContract(TestCase):
    """
    run test with python -m unittest tests.test_owner_count
    """

    HOMESTEAD_BLOCK = 1150000

    def __init__(self, *args, **kwargs):
        super(TestContract, self).__init__(*args, **kwargs)
        self.s = t.state()
        self.s.block.number = self.HOMESTEAD_BLOCK
        t.gas_limit = 4712388

    def test(self):
        # Create 50 accounts
        accounts = []
        ids = []
        keys = []
        account_count = 50
        for i in range(account_count):
            keys.append(sha3(to_string(i)))
            accounts.append(privtoaddr(keys[-1]))
            ids.append(i);
            self.s.block.set_balance(accounts[-1], 10**18)
        # Create wallet
        required_accounts = account_count
        constructor_parameters = (
            accounts,
            ids,
            required_accounts
        )
        self.multisig_wallet = self.s.abi_contract(
            open('solidity/IdentifiedMultiSigWallet.sol').read(),
            language='solidity',
            constructor_parameters=constructor_parameters
        )
        # Create ABIs
        multisig_abi = self.multisig_wallet.translator
        # Send money to wallet contract
        deposit = 1000
        self.s.send(keys[0], self.multisig_wallet.address, deposit)
        self.assertEqual(self.s.block.get_balance(self.multisig_wallet.address), 1000)
        # Transfer money from wallet
        nonce = self.multisig_wallet.getNonce(accounts[0], deposit, "")
        transaction_hash = self.multisig_wallet.submitTransaction(accounts[0], deposit, "", nonce, sender=keys[0])
        include_pending = True
        include_executed = True
        exclude_executed = False
        exclude_pending = False
        self.assertEqual(self.multisig_wallet.getTransactionHashes(0, 1, include_pending, exclude_executed),
                         [transaction_hash])
        for i in range(1, account_count):
            self.multisig_wallet.confirmTransaction(transaction_hash, sender=keys[i])
        # Transaction was executed
        self.assertTrue(self.multisig_wallet.transactions(transaction_hash)[4])
        self.assertEqual(self.multisig_wallet.getTransactionHashes(0, 1, exclude_pending, include_executed),
                         [transaction_hash])
        # Delete owner wa_3. All parties have to confirm.
        remove_owner_data = multisig_abi.encode("removeOwner", [accounts[account_count - 1]])
        nonce = self.multisig_wallet.getNonce(self.multisig_wallet.address, 0, remove_owner_data)
        transaction_hash_2 = self.multisig_wallet.submitTransaction(self.multisig_wallet.address, 0, remove_owner_data,
                                                                    nonce, sender=keys[0])
        self.assertEqual(self.multisig_wallet.getTransactionHashes(0, 1, include_pending, exclude_executed),
                         [transaction_hash_2])
        self.assertEqual(self.multisig_wallet.getTransactionHashes(0, 1, exclude_pending, include_executed),
                         [transaction_hash])
        for i in range(1, account_count):
            profiling = self.multisig_wallet.confirmTransaction(transaction_hash_2, sender=keys[i], profiling=True)
            self.assertLess(profiling['gas'], 100000)
        self.assertEqual(self.multisig_wallet.getTransactionHashes(0, 2, exclude_pending, include_executed),
                         [transaction_hash, transaction_hash_2])
        # Transaction was successfully processed
        self.assertEqual(self.multisig_wallet.required(), required_accounts - 1)

File: contracts/tests/testOwnerCount.js
var async = require('async');
var chaiAsPromised = require('chai-as-promised');
var chai = require('chai');
var expect = chai.expect; // we are using the "expect" style of Chai
chai.use(chaiAsPromised);
var Web3 = require('web3');
var web3 = new Web3();
web3.setProvider(new web3.providers.HttpProvider('http://localhost:8545'));

var wa_1 = 1;
var wa_2 = 2;
var wa_3 = 3;
var wa_4 = 4;
var id_1 = 123456789;
var id_2 = 987654321;
var id_3 = 192837465;
var id_4 = 918273645;
var required_accounts = 2;
var accounts = web3.eth.accounts;

var identifiedMultiSigWallet;
async.series([
  function(callback) {
    describe('Deploy identifiedMultiSigWallet.', function() {
      it('Contract address should not be undefined', function(done) {
        var identifiedMultiSigWalletABI = require('../abi/identifiedMultiSigWalletABI.js')
        var identifiedMultiSigWalletBytecode = require('../bytecode/identifiedMultiSigWalletBytecode.js');
        var _owners = [accounts[wa_1], accounts[wa_2], accounts[wa_3]];
        var _ids = [id_1, id_2, id_3];
        var _required = required_accounts;
        var identifiedMultiSigWalletContract = web3.eth.contract(identifiedMultiSigWalletABI);
        identifiedMultiSigWallet = identifiedMultiSigWalletContract.new(
         _owners,
         _ids,
         _required,
         {
           from: accounts[wa_1],
           data: identifiedMultiSigWalletBytecode,
           gas: '4700000'
         },
         function (e, contract){
            if (typeof contract.address !== 'undefined') {
              done();
              callback();
            }
         });
      });
    });
  },
  function(callback) {
    describe('Sending Money:', function() {
      it('Money should be sent succesfully to the wallet', function() {
        web3.eth.sendTransaction({
          from: accounts[wa_1],
          to: identifiedMultiSigWallet.address,
          value: web3.toWei(1),
          gas: '3000000'
        });
        expect(parseInt(web3.eth.getBalance(identifiedMultiSigWallet.address).toString())).to.equal(parseInt(web3.toWei(1)));
      });
      var excludePending = false;
      var includePending = true;
      var excludeExecuted = false;
      var includeExecuted = true;
      it('Money should be sent succesfully from the wallet', function() {
        let nonce = identifiedMultiSigWallet.getNonce(accounts[0], web3.toWei(1), "");
        identifiedMultiSigWallet.submitTransaction(accounts[0], web3.toWei(1), "", nonce, {
          from: accounts[wa_1],
          to: identifiedMultiSigWallet.address,
          value: 0,
          gas: '3000000'
        }, function(e, txHash) {
          let hash = identifiedMultiSigWallet.getTransactionHashes(0, identifiedMultiSigWallet.getTransactionCount(includePending, includeExecuted), includePending, excludeExecuted)[0];
          identifiedMultiSigWallet.confirmTransaction(hash,{
            from: accounts[wa_2],
            to: identifiedMultiSigWallet.address,
            gas: '3000000'
          }, function(e, res) {
            expect(parseInt(web3.eth.getBalance(identifiedMultiSigWallet.address).toString())).to.equal(0);
            done();
          });
        });
      });
      it(accounts[wa_3] + ' should be removed as owner', function() {
        let nonce = identifiedMultiSigWallet.getNonce(accounts[wa_3]);
        let transactionData = identifiedMultiSigWallet.removeOwner.getData(accounts[wa_3]);
        identifiedMultiSigWallet.submitTransaction(identifiedMultiSigWallet.address, 0, transactionData, nonce, {
          from: accounts[wa_1],
          to: identifiedMultiSigWallet.address,
          value: 0,
          gas: '3000000'
        }, function(e, txHash) {
          let hash = identifiedMultiSigWallet.getTransactionHashes(0, identifiedMultiSigWallet.getTransactionCount(includePending, includeExecuted), includePending, excludeExecuted)[0];
          identifiedMultiSigWallet.confirmTransaction(hash,{
            from: accounts[wa_2],
            to: identifiedMultiSigWallet.address,
            gas: '3000000'
          }, function(e, res) {
            expect(identifiedMultiSigWallet.isOwner(accounts[wa_3])).to.equal(false);
            expect(identifiedMultiSigWallet.ownerWithId(id_3)).to.equal(0);
            expect(identifiedMultiSigWallet.ids(accounts[wa_3])).to.throw(Error);
            done();
          });
        });
      });
    });
  }
]);
